### PR TITLE
p2p: Daniel/gossiping

### PR DESCRIPTION
**Pull request by dziabko**
## Changes
- Added a broadcaster & broadcast handler 
- Added Id interface for Messages
- Implemented gossiping

## Tests:
```
 go test ./p2p -run TestGossipSub
```

### Issues:
close: #320

### CODE REVIEW

File: p2p/message.go [hunk 1]
```diff
@@ -304,8 +315,17 @@ func (bm *BlockAnnounceMessage) Decode(r io.Reader) error {
 	return err
 }
 
+// Id returns the unique Id of the block
+func (bm *BlockAnnounceMessage) Id() string {
+	hash, err := common.Blake2bHash([]byte(bm.String()))
```
Comment (noot) [inline p2p/message.go]: I think this returns the block hash, not the ID of the message?  Could you rename the function to be more clear

File: p2p/message.go [hunk 2]
```diff
@@ -401,6 +426,20 @@ func (tm *TransactionMessage) Decode(r io.Reader) error {
 	return nil
 }
 
+// Id returns the Id of TransactionMessage
+func (tm *TransactionMessage) Id() string {
```
Comment (noot) [inline p2p/message.go]: same here, looks like it's returning the hash

File: p2p/p2p.go [hunk 3]
```diff
@@ -43,19 +43,25 @@ import (
 )
 
 const ProtocolPrefix = "/substrate/dot/2"
+const ProtocolPrefix2 = "/substrate/dot/3"
```
Comment (noot) [inline p2p/p2p.go]: remove this

File: p2p/p2p.go [hunk 4]
```diff
@@ -188,6 +206,57 @@ func (s *Service) Stop() <-chan error {
...
 
+// Broadcast sends a message to all peers
+func (s *Service) Broadcast(msg Message) (err error) {
+	//Get each node it's connected to & broadcast message to them
+
+	//If haven't received the message yet, add to list & rebroadcast it
+	msgType := msg.GetType()
+
+	switch msgType {
+	case StatusMsgType:
```
Comment (noot) [inline p2p/p2p.go]: we don't need to broadcast status messages

File: p2p/p2p.go [hunk 5]
```diff
@@ -256,12 +355,6 @@ func (s *Service) Ctx() context.Context {
 	return s.ctx
 }
 
-// PeerCount returns the number of connected peers
```
Comment (noot) [inline p2p/p2p.go]: why is this removed?

File: p2p/p2p.go [hunk 6]
```diff
@@ -120,7 +131,14 @@ func NewService(conf *Config, msgChan chan<- Message) (*Service, error) {
...
 	}
 
+	s.statusMessagesRec = make(map[string]bool)
+	s.blockReqRec = make(map[string]bool)
+	s.blockRespRec = make(map[string]bool)
+	s.blockAnnounceRec = make(map[string]bool)
+	s.txMessageRec = make(map[string]bool)
+
 	h.SetStreamHandler(ProtocolPrefix, s.handleStream)
+	h.SetStreamHandler(ProtocolPrefix2, s.handleBroadcastStream)
```
Comment (noot) [inline p2p/p2p.go]: there is only one protocol prefix for a network, and there should only be one stream handler for it

File: p2p/p2p.go [hunk 7]
```diff
@@ -378,6 +471,61 @@ func (s *Service) Peers() []string {
...
+	// decode message
+	msg, err := DecodeMessage(rw.Reader)
+	if err != nil {
+		log.Error("failed to decode message", "error", err)
+		return
+	}
+
+	log.Debug("got message", "peer", stream.Conn().RemotePeer(), "type", msgType, "msg", msg.String())
+
+	err = s.Broadcast(msg)
```
Comment (noot) [inline p2p/p2p.go]: not sure what the difference in between this and the other handle stream function apart from the broadcasting, you should remove this and add logic to handle broadcasting for different message types inside the existing stream handler

File: p2p/p2p_test.go [hunk 8]
```diff
@@ -186,3 +189,109 @@ func TestSend(t *testing.T) {
 		t.Errorf("Send error: %s", err)
 	}
 }
+
+func TestGossipSub(t *testing.T) {
```
Comment (noot) [inline p2p/p2p_test.go]: not sub :p lol

File: p2p/p2p.go [hunk 9]
```diff
@@ -43,19 +43,25 @@ import (
...
+	ctx               context.Context
+	host              core.Host
+	hostAddr          ma.Multiaddr
+	dht               *kaddht.IpfsDHT
+	dhtConfig         kaddht.BootstrapConfig
+	bootstrapNodes    []peer.AddrInfo
+	mdns              discovery.Service
+	msgChan           chan<- Message
+	noBootstrap       bool
+	statusMessagesRec map[string]bool
```
Comment (noot) [inline p2p/p2p.go]: can probably make these `map[common.Hash]bool` since we are hashing the message to identify it

File: p2p/message.go [hunk 10]
```diff
@@ -401,6 +426,20 @@ func (tm *TransactionMessage) Decode(r io.Reader) error {
...
+
+	var hashedExtrinsics = make([]byte, 0)
+	for _, extrinsic := range tm.Extrinsics {
+		hashedExt, err := common.Blake2b128([]byte(extrinsic))
+		if err != nil {
+			return ""
+		}
+		hashedExtrinsics = append(hashedExtrinsics, hashedExt...)
+	}
+	return string(hashedExtrinsics)
```
Comment (noot) [inline p2p/message.go]: we should just encode then hash the entire message and return a common.Hash

File: p2p/message.go [hunk 11]
```diff
@@ -304,8 +315,17 @@ func (bm *BlockAnnounceMessage) Decode(r io.Reader) error {
 	return err
 }
 
+// Id returns the unique Id of the block
+func (bm *BlockAnnounceMessage) Id() string {
+	hash, err := common.Blake2bHash([]byte(bm.String()))
+	if err != nil {
+		return ""
+	}
+	return hash.String()
```
Comment (noot) [inline p2p/message.go]: we should just encode then hash the entire message and return a common.Hash

File: p2p/message.go [hunk 12]
```diff
@@ -273,6 +279,11 @@ func (bm *BlockRequestMessage) Decode(r io.Reader) error {
 	return nil
 }
 
+// Id Returns the ID of the block
+func (bm *BlockRequestMessage) Id() string {
+	return string(bm.ID)
```
Comment (noot) [inline p2p/message.go]: I think we should also probably encode+hash the entire message for BlockRequest and BlockResponse as well, since there's a chance of people using duplicate IDs

**Comment by noot** [comment]
nice work :) please fix conflicts as well!

**Comment by noot** [comment]
also could you update the description to say what issue this closes: https://github.com/ChainSafe/gossamer/issues/269

File: p2p/message.go (hunk 12 above)
Comment (dziabko) [reply p2p/message.go]: We're assuming the ID is a random number that is unique

File: p2p/message.go [hunk 13]
```diff
@@ -127,10 +128,15 @@ func (sm *StatusMessage) Decode(r io.Reader) error {
 	return err
 }
 
+// Decodes the message into a StatusMessage, it assumes the type byte has been removed
```
Comment (ansermino) [inline p2p/message.go]: Update comment

File: p2p/p2p.go [hunk 14]
```diff
@@ -68,6 +72,11 @@ type Config struct {
 	NoMdns         bool
 }
 
+type MessageWithID struct {
```
Comment (ansermino) [inline p2p/p2p.go]: What is this?

File: p2p/p2p.go [hunk 15]
```diff
@@ -189,6 +203,52 @@ func (s *Service) Stop() <-chan error {
 	return e
 }
 
+// Broadcast sends a message to all peers
+func (s *Service) Broadcast(msg Message) (err error) {
+	//Get each node it's connected to & broadcast message to them
+
+	//If haven't received the message yet, add to list & rebroadcast it
```
Comment (ansermino) [inline p2p/p2p.go]: Please cleanup & clarify these comments

File: p2p/p2p.go [hunk 16]
```diff
@@ -220,6 +280,36 @@ func (s *Service) Send(peer core.PeerAddrInfo, msg []byte) (err error) {
 	return nil
 }
 
+// SendBroadcast streams to broadcast stream handler
+func (s *Service) SendBroadcast(peer core.PeerAddrInfo, msg []byte) error {
```
Comment (ansermino) [inline p2p/p2p.go]: Wha'ts different between this and `Send`?

File: p2p/p2p_test.go [hunk 17]
```diff
@@ -214,3 +214,97 @@ func TestSend(t *testing.T) {
...
+		StartingBlock: []byte{1, 1},
+		EndBlockHash:  optional.NewHash(true, endBlock),
+		Direction:     1,
+		Max:           optional.NewUint32(true, 1),
+	}
+
+	nodeA.Broadcast(bm)
+
+	select {
+	case <-msgChanB:
```
Comment (ansermino) [inline p2p/p2p_test.go]: Please validate the contents of the msg

File: p2p/p2p_test.go [hunk 18]
```diff
@@ -214,3 +214,97 @@ func TestSend(t *testing.T) {
...
+	defer nodeC.Stop()
+
+	//Connect node B & node C
+	err = nodeC.bootstrapConnect()
+	if err != nil {
+		t.Errorf("Start error :%s", err)
+	}
+
+	// Create mock BlockRequestMessage to broadcast
+	endBlock, err := common.HexToHash("0xfd19d9ebac759c993fd2e05a1cff9e757d8741c2704c8682c15b5503496b6aa1")
```
Comment (ansermino) [inline p2p/p2p_test.go]: Not a fan of magic values. Please document what input you used for this or just hash a concrete instance

File: p2p/message.go (hunk 12 above)
Comment (noot) [reply p2p/message.go]: okay, I'm still unsure about this since I think the chance of duplicate IDs is pretty high. I guess this is ok for now but we should probably change it to the hash of the entire message as well, or clear the cache of seen messages after some time
